integ/src/monitor.rs

```diff
@@ -0,0 +1,533 @@
+    // fetch brupop resources - Controllers, Agents, Apiserver to help on determining if they are on ideal status.
+    async fn fetch_brupop_resources(&self) -> Result<ObjectList<Pod>>;
+}
+
+#[derive(Clone)]
+/// Concrete implementation of the `BrupopClient` trait. This implementation will almost
+/// certainly be used in any case that isn't a unit test.
+pub struct IntegBrupopClient {
+    k8s_client: kube::client::Client,
+}
+
+impl IntegBrupopClient {
+    pub fn new(k8s_client: kube::client::Client) -> Self {
+        IntegBrupopClient { k8s_client }
+    }
+}
+
+#[async_trait]
+impl BrupopClient for IntegBrupopClient {
+    async fn fetch_shadows(&self) -> Result<ObjectList<BottlerocketShadow>> {
+        let brss: Api<BottlerocketShadow> = Api::namespaced(self.k8s_client.clone(), NAMESPACE);
+        let brss_object_list = brss
+            .list(&ListParams::default())
+            .await
+            .context(monitor_error::FindBrupopResources {})?;
+
+        Ok(brss_object_list)
+    }
+
+    async fn fetch_brupop_resources(&self) -> Result<ObjectList<Pod>> {
+        let pods: Api<Pod> = Api::namespaced(self.k8s_client.clone(), NAMESPACE);
+        let pods_objectlist = pods
+            .list(&ListParams::default())
+            .await
+            .context(monitor_error::FindBrupopResources {})?;
+
+        Ok(pods_objectlist)
+    }
+}
+
+#[async_trait]
+/// A trait providing an interface to monitor brupop process.
+pub trait Monitor: Clone {
+    async fn run_monitor(&self) -> Result<()>;
+}
+
+#[derive(Clone)]
+pub struct BrupopMonitor<T: BrupopClient> {
+    integ_brupop_client: T,
+}
+
+impl<T: BrupopClient> BrupopMonitor<T> {
+    pub fn new(integ_brupop_client: T) -> Self {
+        BrupopMonitor {
+            integ_brupop_client,
+        }
+    }
+
+    // verify if Burpop resources (agent, api-server, controller) are in `running` status.
+    async fn check_resources_health(&self, pods: &ObjectList<Pod>) -> Result<bool> {
+        for pod in pods {
+            if !is_pod_running(&pod) {
+                return Ok(false);
+            }
+        }
+
+        Ok(true)
+    }
+
+    // verify if brs has been created properly and initialized `status`.
+    async fn check_shadows_health(
+        &self,
+        bottlerocketshadows: &ObjectList<BottlerocketShadow>,
+    ) -> Result<bool> {
+        for bottlerocketshadow in bottlerocketshadows {
+            // verify if brs has been updated `status` properly
+            if bottlerocketshadow.status.is_some() {
+                continue;
+            } else {
+                return Ok(false);
+            }
+        }
+
+        Ok(true)
+    }
+
+    // confirm that the instances successfully made it to the target version and the Idle state
+    async fn confirm_update_success(
+        &self,
+        bottlerocketshadows: &ObjectList<BottlerocketShadow>,
+    ) -> Result<bool> {
+        let mut update_success = 1;
```

Comment:
This seems to try to use int to do bool's job. Maybe use bool would be more clear? like
```
let mut update_success = true;

if ...


{
   update_success = update_success & false;
}

Ok(update_success)
```